tests/regression/test_project_interp_KMV.py

```diff
@@ -0,0 +1,67 @@
+import pytest
+import numpy as np
+from firedrake import *
+import FIAT
+import finat
+
+
+def run_interpolation_KMV(MeshClass, r, d):
+    if MeshClass.__name__ == "UnitSquareMesh":
+        mesh = MeshClass(2 ** r, 2 ** r)
+        x, y = SpatialCoordinate(mesh)
+        f = (1 + 8 * pi * pi) * cos(x * pi * 2) * cos(y * pi * 2)
+    elif MeshClass.__name__ == "UnitCubeMesh":
+        mesh = MeshClass(2 ** r, 2 ** r, 2 ** r)
+        x, y, z = SpatialCoordinate(mesh)
+        f = (1 + 8 * pi * pi) * cos(x * pi * 2) * cos(y * pi * 2) * cos(z * pi * 2)
```

Comment:
Please use fixtures for this.

```python
@pytest.fixture(params=["square", "cube"])
def mesh_type(request):
    return request.param


@pytest.fixture
def mesh(mesh_type):
    return {"square": lambda n: UnitSquareMesh(n, n),
                "cube": lambda n: UnitCubeMesh(n, n, n)}[mesh_type]


@pytest.fixture
def max_degree(mesh_type):
    return {"square": 6, "cube": 4}[mesh_type]


@pytest.fixture
def interpolation_expr(mesh_type):
    return {"square": lambda x, y: (1 + 8 * pi * pi) * cos(x * pi * 2) * cos(y * pi * 2),
                "cube": lambda x, y, z: (1 + 8 * pi * pi) * cos(x * pi * 2) * cos(y * pi * 2) * cos(z * pi * 2)}[mesh_type]


def run_interpolation(mesh, expr, p):
    expr = expr(*SpatialCoordinate(mesh))
    V = FunctionSpace(mesh, "KMV", p)
    return errornorm(expr, interpolate(expr, V))


def test_interpolation_KMV(mesh, max_degree, interpolation_expr):
     for p in range(1, max_degree):
         errors = [run_interpolation(mesh(r), interpolation_expr, p) for r in range(3, 6)]
         ...
```

The same can be done for the projection.

Reply:
Sure thing. Just out of curiosity, why would one prefer `fixtures` instead of `parameterizations` here?

Reply:
You move your "business-logic" out of the test.

Reply:
I like it. It's quite a bit neater than parameterizations.

Reply:
The other thing you can have is that fixtures can depend on each other (rather than having to manually create pairs and then destructure them in the tests). But be wary of accidentally taking cartesian products of test parameters.

Reply:
Okay, I revamped the tests to reflect your comments.